### PR TITLE
Optionally serializing ubyte[] as base64 string

### PR DESCRIPTION
**Pull request by ferencdg**
This can save brandwidth and decrease latency of REST queries

usage:
```D
@serPolicy!(Base64ArraySerPolicy)
interface ITestBase64
{
    @safe X getTest();
}
```

### CODE REVIEW

**Comment by ferencdg** [comment]
this is an other serialization change
@s-ludwig @Geod24 please let me know if anything should be changed

File: data/vibe/data/serialization.d [hunk 1]
```diff
@@ -1487,6 +1488,22 @@ private template getExpandedFieldsData(T, FIELDS...)
...
+		import std.base64 : Base64;
+		return Base64.encode(data);
+	}
+
+	public static ubyte[] fromRepresentation (in string data) @safe pure
+	{
+		import std.base64 : Base64;
+		return Base64.decode(data);
+	}
+}
```
Comment (Geod24) [inline data/vibe/data/serialization.d:1505]: So, should we try writing to a sink / input range here ? :D

**Comment by s-ludwig** [comment]
I wonder why the policy logic needs to be repeated in the web framework. Couldn't you just use the `@resultSerializer` attribute to specify a serializer that applies a certain policy?

File: data/vibe/data/serialization.d (hunk 1 above)
Comment (ferencdg) [reply data/vibe/data/serialization.d:1505]: I think if we want to have the policies support sink type serialization, then it should be a separate ticket. Polices were added to vibe.d long time ago.

**Comment by ferencdg** [comment]
> I wonder why the policy logic needs to be repeated in the web framework. Couldn't you just use the `@resultSerializer` attribute to specify a serializer that applies a certain policy?
 
1. I wanted to have an easy way to specify the policy for ALL interface methods, so I declared it on the interface. ResultSerializers are attached to interface methods.
2. Result serializer only serializes the results, but when we define a policy on the interface it will affect the serialization of everything: result, query parameters, headers...

**Comment by ferencdg** [comment]
> > I wonder why the policy logic needs to be repeated in the web framework. Couldn't you just use the `@resultSerializer` attribute to specify a serializer that applies a certain policy?
> 
> 1. I wanted to have an easy way to specify the policy for ALL interface methods, so I declared it on the interface. ResultSerializers are attached to interface methods, I would find it tedious to create REST interfaces and attach resultSerializer to all methods just to add a new policy.
> 2. Result serializer only serializes the results, but when we define a policy on the interface it will affect the serialization of everything: result, query parameters, headers...

@s-ludwig please let me know if you would like to see a different design

File: web/vibe/web/common.d [hunk 2]
```diff
@@ -583,6 +584,88 @@ package template ResultSerializersT(alias func) {
...
+}
+
+///
+package struct SerPolicy (alias PolicyTemplatePar)
+{
+	alias PolicyTemplate = PolicyTemplatePar;
+}
+
+///
+public alias serPolicy (Args...) = SerPolicy!(Args);
```
Comment (Geod24) [inline web/vibe/web/common.d:607]:
```suggestion
public alias serializationPolicy (Args...) = SerializationPolicy!(Args);
```

It's better to be a little more verbose in the name for this public-facing method.

**Review by Geod24 — APPROVED** [review]
LGTM, just the one nit on the name, and I think this could use some documentation.

**Comment by Geod24** [comment]
I did a few modifications:
- Fixed the documentation (indent was wrong, and the section name was wrong too, it should have been `Serialization_Policies:` instead of `Serialization Policies:` to count as a section name);
- Renamed `Base64ArraySerPolicy` to simply `Base64ArrayPolicy`, since it's already in the serialization package;
- Slightly rephrased the documentation and extended it to say it applies to *all* types;

@ferencdg : Sorry I missed it earlier, but why is this only on return values ? It should also apply to parameters. In fact, the issue was noticed on a parameter (when accepting an `SCPEnvelope`).

**Review by Geod24 — APPROVED** [review]

**Pull request merged**